### PR TITLE
feat(resource): add meta service resource

### PR DESCRIPTION
**Pull request by adr-mo**
This PR intends to handle metaservices in the resource status view.

To do so it
- Adds Meta Service to resource status listing
- Adds Meta Service detail to resource status
- Adds Meta Service graph to resource status
- Adds Meta Service Shortcuts to resource status
- Adds monitoring actions: ack, dt, submit result, checks and comment for metaservices

It brings new endpoints
- GET `/monitoring/metaservices/{metaId}/metrics` - list metrics of the metaservice (realtime)
- GET `/configuration/metaservices` - list metaservices (configuration)
- GET `/configuration/metaservices/{metaId}` - gets metaservice configuration

Other
- Refacto of the ResourceMonitoring repository using symfony providers

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [ ] 20.04.x
- [ ] 20.10.x
- [x] 21.04.x (master)

<h2> How this pull request can be tested ? </h2>

See Jira ticket for tests

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 1]
```diff
@@ -16090,6 +16090,20 @@ msgstr "Erreur lors de l'ajout d'une catégorie hôte"
...
+msgstr "Erreur lors de la recherche des configurations des meta services"
+
+msgid "Error when searching for the meta service configuration (%s)"
+msgstr "Erreur lors de la recherche du meta service (%s)"
+
+msgid "Calculation method provided not supported (%s)"
+msgstr "Méthode de calcul fournie non supportée (%s)"
+
+msgid "Meta Service linked to service %d not found"
+msgstr "Le Meta Service lié au service %d non trouvé"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:16103]:
```suggestion
msgstr "Meta Service lié au service %d non trouvé"
```

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 2]
```diff
@@ -16090,6 +16090,20 @@ msgstr "Erreur lors de l'ajout d'une catégorie hôte"
...
+
+msgid "Host meta for meta %d not found"
+msgstr "L'hôte lié au meta %d non trouvé"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:16106]:
```suggestion
msgstr "Hôte lié au meta %d non trouvé"
```

File: src/Centreon/Application/Controller/Monitoring/TimelineController.php [hunk 3]
```diff
@@ -157,4 +157,59 @@ public function getServiceTimeline(
...
+        $user = $this->getUser();
+        $this->monitoringService->filterByContact($user);
+        $this->timelineService->filterByContact($user);
+
+        $service = $this->monitoringService->findOneServiceByDescription('meta_' . $metaId);
+
+        if (is_null($service)) {
+            throw new EntityNotFoundException(
+                sprintf(
+                    _('Meta %d not found'),
```
Comment (kduret) [inline src/Centreon/Application/Controller/Monitoring/TimelineController.php:187]:
```suggestion
                    _('Meta service %d not found'),
```

File: src/Centreon/Application/Controller/Monitoring/CommentController.php [hunk 4]
```diff
@@ -286,4 +302,59 @@ public function addServiceComment(
...
+         * At this point we validate the JSON sent with the JSON validator.
+         */
+        $date = ($receivedData['date'] !== null) ? new \DateTime($receivedData['date']) : new \DateTime();
+
+        $service = $this->monitoringService->findOneServiceByDescription('meta_' . $metaId);
+
+        if (is_null($service)) {
+            throw new EntityNotFoundException(
+                sprintf(
+                    _('Meta %d not found'),
```
Comment (kduret) [inline src/Centreon/Application/Controller/Monitoring/CommentController.php:346]:
```suggestion
                    _('Meta service %d not found'),
```

File: src/Centreon/Application/Controller/Monitoring/TimelineController.php [hunk 5]
```diff
@@ -157,4 +157,59 @@ public function getServiceTimeline(
...
+                    $metaId
+                )
+            );
+        }
+
+        $host = $this->monitoringService->findOneHost($service->getHost()->getId());
+        if (is_null($host)) {
+            throw new EntityNotFoundException(
+                sprintf(_('Host meta for meta %d not found'), $metaId)
```
Comment (kduret) [inline src/Centreon/Application/Controller/Monitoring/TimelineController.php:196]:
```suggestion
                sprintf(_('Host meta for meta service %d not found'), $metaId)
```

File: src/Centreon/Domain/MetaServiceConfiguration/Interfaces/MetaServiceConfigurationReadRepositoryInterface.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    /**
+     * Find all meta services configurations.
+     *
+     * @return MetaServiceConfiguration[]
+     * @throws \Throwable
+     */
+    public function findAll(): array;
+
+    /**
+     * Find all meta services configurations. by contact.
```
Comment (kduret) [inline src/Centreon/Domain/MetaServiceConfiguration/Interfaces/MetaServiceConfigurationReadRepositoryInterface.php:63]:
```suggestion
     * Find all meta services configurations by contact.
```

File: src/Centreon/Domain/Monitoring/MetaService/UseCase/V21/MetaServiceMetric/FindMetaServiceMetrics.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace Centreon\Domain\Monitoring\MetaService\UseCase\V21\MetaServiceMetric;
+
+use Centreon\Domain\Contact\Interfaces\ContactInterface;
+use Centreon\Domain\Monitoring\MetaService\Exception\MetaServiceMetricException;
+use Centreon\Domain\Monitoring\MetaService\UseCase\V21\MetaServiceMetric\FindMetaServiceMetricsResponse;
+use Centreon\Domain\Monitoring\MetaService\Interfaces\MetaServiceMetric\MetaServiceMetricServiceInterface;
+
+/**
+ * This class is designed to represent a use case to find all host categories.
```
Comment (kduret) [inline src/Centreon/Domain/Monitoring/MetaService/UseCase/V21/MetaServiceMetric/FindMetaServiceMetrics.php:31]:
```suggestion
 * This class is designed to represent a use case to find all meta services metrics.
```

File: src/Centreon/Infrastructure/MetaServiceConfiguration/Repository/MetaServiceConfigurationRepositoryRDB.php [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+                        WHEN ms.calcul_type = 'AVE' THEN 'average'
+                        WHEN ms.calcul_type = 'SOM' THEN 'sum'
+                        WHEN ms.calcul_type = 'MIN' THEN 'minimum'
+                        WHEN ms.calcul_type = 'MAX' THEN 'maximum'
+                    END AS `calculation_type`
+                FROM `:db`.meta_service ms"
+            );
+        } else {
+            $request = $this->translateDbName(
+                "SELECT SQL_CALC_FOUND_ROWS
```
Comment (kduret) [inline src/Centreon/Infrastructure/MetaServiceConfiguration/Repository/MetaServiceConfigurationRepositoryRDB.php:139]: `SELECT` could be construct before `if` statement
Then, if contact id is set, add `JOIN`

File: src/Centreon/Infrastructure/MetaServiceConfiguration/Repository/MetaServiceConfigurationRepositoryRDB.php [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+     * @param int|null $contactId Contact id related to host categories
+     * @return MetaServiceConfiguration|null
+     * @throws AssertionFailedException
+     */
+    private function findByIdRequest(int $metaId, ?int $contactId): ?MetaServiceConfiguration
+    {
+        if ($contactId === null) {
+            $statement = $this->db->prepare(
+                $this->translateDbName(
+                    "SELECT ms.meta_id,
```
Comment (kduret) [inline src/Centreon/Infrastructure/MetaServiceConfiguration/Repository/MetaServiceConfigurationRepositoryRDB.php:250]: `SELECT` could be construct before `if` statement
Then, if contact id is set, add `JOIN`

File: tests/php/Centreon/Infrastructure/MetaServiceConfiguration/Repository/Model/MetaServiceConfigurationFactoryRdbTest.php [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+declare(strict_types=1);
+
+namespace Tests\Centreon\Infrastructure\MetaServiceConfiguration\Repository\Model;
+
+use Centreon\Infrastructure\MetaServiceConfiguration\Repository\Model\MetaServiceConfigurationFactoryRdb;
+
```
Comment (kduret) [inline tests/php/Centreon/Infrastructure/MetaServiceConfiguration/Repository/Model/MetaServiceConfigurationFactoryRdbTest.php:26]:
```suggestion
```

File: src/Centreon/Infrastructure/MetaServiceConfiguration/API/Model/MetaServiceConfigurationV21Factory.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
+<?php
```
Comment (jeremyjaouen) [inline src/Centreon/Infrastructure/MetaServiceConfiguration/API/Model/MetaServiceConfigurationV21Factory.php:1]: I think you could look at : https://github.com/centreon/centreon-autodiscovery/pull/434/files#diff-54be133dcd0f313d03c987c66a05279fe8e38a9011bda065d708dda6688baa4b. This way to do your factory could avoid you to have a create your MetaServiceConfigurationV21.php file, and directly create your class in your factory.

File: src/Centreon/Infrastructure/Monitoring/MetaService/API/Model/MetaServiceMetricFactoryV21.php [hunk 12]
```diff
@@ -0,0 +1,55 @@
+<?php
```
Comment (jeremyjaouen) [inline src/Centreon/Infrastructure/Monitoring/MetaService/API/Model/MetaServiceMetricFactoryV21.php:1]: I think you could look at : https://github.com/centreon/centreon-autodiscovery/pull/434/files#diff-54be133dcd0f313d03c987c66a05279fe8e38a9011bda065d708dda6688baa4b. This way to do your factory could avoid you to have a create your MetaServiceMetricV21.php file, and directly create your class in your factory.

File: tests/php/Centreon/Infrastructure/MetaServiceConfiguration/Repository/Model/MetaServiceConfigurationFactoryRdbTest.php (hunk 10 above)
Comment (adr-mo) [reply tests/php/Centreon/Infrastructure/MetaServiceConfiguration/Repository/Model/MetaServiceConfigurationFactoryRdbTest.php]: done

File: config/routes/Centreon/monitoring/metric.yaml [hunk 13]
```diff
@@ -37,3 +37,17 @@ monitoring.metric.getServiceStatusMetrics:
         serviceId: '\d+'
     controller: 'Centreon\Application\Controller\Monitoring\MetricController::getServiceStatusMetrics'
     condition: "request.attributes.get('version') >= 2.0"
+
```
[the remark this answers is not in the document]
Comment (adr-mo) [reply config/routes/Centreon/monitoring/metric.yaml:40]: IMHO I think we could do this during the refacto of the resource status backend. I've followed what was done previously to ease development for now.

File: src/Centreon/Infrastructure/MetaServiceConfiguration/API/Model/MetaServiceConfigurationV21Factory.php [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+            $newMetaServiceConfiguration->metric = $metaServiceConfiguration['metric'];
+            $newMetaServiceConfiguration->warning = $metaServiceConfiguration['warning'];
+            $newMetaServiceConfiguration->critical = $metaServiceConfiguration['critical'];
+
+            $metaServicesConfigurations[] = $newMetaServiceConfiguration;
+        }
+        return $metaServicesConfigurations;
+    }
+
+    private static function createEmptyClass(): \stdClass
```
Comment (kduret) [inline src/Centreon/Infrastructure/MetaServiceConfiguration/API/Model/MetaServiceConfigurationV21Factory.php:85]:
```suggestion
    /**
     * @return \stdClass
     */
    private static function createEmptyClass(): \stdClass
```

**Review by kduret — DISMISSED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

**Review by kduret — APPROVED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Pull request merged**